### PR TITLE
Fixing hpx::wait_all() invoked with two vector<future<T>> 

### LINKED ISSUE #1623: hpx::wait_all() invoked with two vector<future<T>> fails
This code snippet currently fails to compile (at least with MSVC):

    std::vector<hpx::future<int> > v1 = { ... };
    std::vector<fhpx::uture<int> > v2 = { ... };
    hpx::wait_all(v1, v2);

### PR DESCRIPTION
**Pull request by hkaiser**
This fixes #1623

### CODE REVIEW

File: hpx/lcos/detail/async_implementations.hpp [hunk 1]
```diff
@@ -148,11 +149,11 @@ namespace hpx { namespace detail
 
         if (target_is_managed)
         {
-            typedef typename lcos::detail::shared_state_ptr_for<
+            typedef typename traits::detail::shared_state_ptr_for<
                 future<result_type>
             >::type shared_state_ptr;
 
-            shared_state_ptr const& state = lcos::detail::get_shared_state(f);
+            shared_state_ptr const& state = traits::get_shared_state(f);
```
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: This looks odd. The trait is defined in `hpx::traits` but the returned type is defined in `hpx::traits::detail`.

File: hpx/lcos/detail/async_implementations.hpp [hunk 2]
```diff
@@ -210,11 +211,11 @@ namespace hpx { namespace detail
...
```
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: Same here.

File: hpx/lcos/wait_any.hpp [hunk 3]
```diff
@@ -181,9 +181,7 @@ namespace hpx { namespace lcos
     }
 
     template <typename Iterator>
-    typename util::always_void<
-        typename lcos::detail::future_iterator_traits<Iterator>::type
-    >::type
+    typename lcos::detail::future_iterator_traits<Iterator>::type
```
Comment (sithhell) [inline hpx/lcos/wait_any.hpp]: This is changing the functions return type, is this intended?

File: hpx/lcos/wait_all.hpp [hunk 4]
```diff
@@ -331,14 +333,12 @@ namespace hpx { namespace lcos
     }
 
     template <typename Iterator>
-    typename util::always_void<
-        typename lcos::detail::future_iterator_traits<Iterator>::type
-    >::type
+    typename lcos::detail::future_iterator_traits<Iterator>::type
```
Comment (sithhell) [inline hpx/lcos/wait_all.hpp]: This is changing the functions return type, is this intended?

File: hpx/lcos/wait_some.hpp [hunk 5]
```diff
@@ -368,16 +388,15 @@ namespace hpx { namespace lcos
     }
 
     template <typename Iterator>
-    typename util::always_void<
-        typename lcos::detail::future_iterator_traits<Iterator>::type
-    >::type wait_some(std::size_t n, Iterator begin, Iterator end,
+    typename lcos::detail::future_iterator_traits<Iterator>::type
```
Comment (sithhell) [inline hpx/lcos/wait_some.hpp]: This is changing the functions return type, is this intended?

File: hpx/lcos/wait_any.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/lcos/wait_any.hpp]: Definitely not intended, excellent catch!

File: hpx/lcos/detail/async_implementations.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/lcos/detail/async_implementations.hpp]: Good point, everything belongs into `traits::detail`.

**Pull request merged**